### PR TITLE
bpo-35226: Fix equality for nested unittest.mock.call objects.

### PR DESCRIPTION
**Pull request by cjw296**
https://bugs.python.org/issue35226

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -2054,6 +2054,10 @@ def __eq__(self, other):
         else:
             self_name, self_args, self_kwargs = self
 
+        if (getattr(self, 'parent', None) and getattr(other, 'parent', None)
```
Comment (mariocj89) [inline Lib/unittest/mock.py:2062]: `self` should always have a parent attribute, no need to use getattr, isn't it?

You are using gettattr to ignore the value and then compare it. What if it has a parent but it evaluates Falsy? Might be better to just  use `hasattr`.

```python
if hasattr(other, 'parent') and self.parent != other.parent:
    return False
```

I think it reads easier

File: Lib/unittest/test/testmock/testhelpers.py [hunk 2]
```diff
@@ -269,6 +269,14 @@ def test_extended_call(self):
         self.assertEqual(mock.mock_calls, last_call.call_list())
 
 
+    def test_extended_not_equal(self):
+        a = call(x=1).foo
```
Comment (mariocj89) [inline Lib/unittest/test/testmock/testhelpers.py:273]: I know it works due to the implementation, but I'd also validate this works recursively for future changes to the module with 
```python 
a = call(x=1).foo().bar
b = call(x=2).foo().bar
```

File: Lib/unittest/mock.py (hunk 1 above)
Comment (cjw296) [reply Lib/unittest/mock.py:2062]: The implementation was suggested in the bug discussion, @voidspace can make a call here.

Comment (voidspace) [reply Lib/unittest/mock.py:2062]: I think @mariocj89 has a point. On `call()` the parent is `None`.

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: Is `hasattr` safe to use here or does it still risk swallowing exceptions other than `AttributeError` on Python 3.6+? The grumpy old python dev in me still prefers `getattr(other, 'parent', None)` out of paranoia...

Comment (voidspace) [reply Lib/unittest/mock.py:2062]: `hasattr` was fixed quite some time ago I believe :-)

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: Okay, but is there benefit in keeping the `getattr(other, 'parent', None)` so that this patch can be identical on https://github.com/testing-cabal/mock?

Comment (cjw296) [reply Lib/unittest/mock.py]: Interestingly, bunch of tests fail after the changes suggested here...

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: @mariocj89 / @voidspace - I'm afraid with the suggested changes, we get a bunch of failure:
https://travis-ci.org/python/cpython/jobs/455556002#L2849-L3008
I don't *think* I've done anything wrong here. I played around locally, and with anything other than exactly as it was originally, you get at least one test failure...

Comment (voidspace) [reply Lib/unittest/mock.py:2062]: Wow, that's unexpected :-)

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: Indeed! @voidspace - are you happy to go with the original solution so we can get this merged?

Comment (mariocj89) [reply Lib/unittest/mock.py:2062]: Surprising indeed!

Look at this:
```python
>>> from unittest.mock import Mock
>>> m = Mock()
>>> m.abc.cde()
<Mock name='mock.abc.cde()' id='139944355306408'>
>>> m.mock_calls
[call.abc.cde()]
>>> print(m.mock_calls[0].parent)
None
```

Seems parent is not set at all, which is causing a lot of tests to fail with my proposal :/

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: Wow. That seems like a bug too, let me add some cases and see about  a fix.
@voidspace - open question: shall we keep `getattr(other, 'parent', None)` to cover the porting-to-testing-cabal version issue?

Comment (mariocj89) [reply Lib/unittest/mock.py:2062]: (just as a note, we can follow this up indeed in another issue as I am not sure about the fix, seems we are not passing the parent here): https://github.com/python/cpython/blob/master/Lib/unittest/mock.py#L986)

Comment (cjw296) [reply Lib/unittest/mock.py:2062]: I'm afraid it's intrinsically linked to this issue, `call` children are designed to be compared to items in `Mock().method_calls` and `Mock().mock_calls` and I don't think that currently works safely... They'll generally always compare as equal, which isn't what we want with nested/generative calls...

**Comment by cjw296** [comment]
@mariocj89 - okay, so I finally got a chance to come back to this.
So, the reason that calls found in `mock_calls` don't have parents set is that the process that builds them is fundamentally different. I've added some tests in c704bca to verify that they don't have the issue I was worried they might have.

@voidspace - I'm happy with this PR now, can we go for a merge? Then I'll get the backporting shenanigans done.

**Comment by cjw296** [comment]
Gah, finally found the case that was the problem :-(

**Comment by cjw296** [comment]
@mariocj89 / @voidspace - okay, after a good few hours of tinkering and thinkering, I think this PR is good to go.

State of play is as follows:

- This PR does fix the issue with `unittest.mock.call`.
- It does *not* fix the issue with `mock_calls`, but does add warnings about the problem to the docs.

I don't believe the `mock_calls` issue can be solved in a backwards compatible way, as a `return_value` that is a mock is not computed per call, but only once. This means that when we come to record `mock_calls` on a nested child, we have no way of knowing what parameters were passed to the ancestor call that created it, and we can't start tracking that without fundamentally changing how `return_value` works in a backwards-incompatible way.

So, if you could review this PR on that basis, and add backport tags to all maintained versions since this is now just a bugfix change, I'd be very grateful.

My plan is to start building a `CallTrackingMock` based on `MagicMock` over in https://github.com/Simplistix/testfixtures. Once I'm comfortable that's a good idea, I'll likely do a feature request PR into cpython. If you have thoughts, I'd welcome comments over on https://github.com/Simplistix/testfixtures/issues/90.

**Comment by cjw296** [comment]
@mariocj89 - if you're okay with this, can you approve it so it gets merged? I can then backport to https://github.com/testing-cabal/mock where I'm now a maintainer.

File: Misc/NEWS.d/next/Library/2018-11-15-07-14-32.bpo-35226.wJPEEe.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix equality for nested unittest.mock.call objects.
```
Comment (mariocj89) [inline Misc/NEWS.d/next/Library/2018-11-15-07-14-32.bpo-35226.wJPEEe.rst]: I'd try to be more explicit here. 

```Recursively check  arguments when testing for equality of :class:`unittest.mock.call` objects.```.
As the parent names are already being checked via the name attribute.

File: Lib/unittest/mock.py [hunk 4]
```diff
@@ -2054,6 +2054,10 @@ def __eq__(self, other):
...
+            and self.parent != other.parent):
```
Comment (mariocj89) [inline Lib/unittest/mock.py]: this needs one more level of indentation

File: Lib/unittest/test/testmock/testmock.py [hunk 5]
```diff
@@ -925,6 +925,50 @@ def test_mock_calls(self):
...
+        # parent looks like this:
+        self.assertEqual(m.mock_calls, [call(), call().wibble()])
+        # but child should look like this:
+        self.assertEqual(result.mock_calls, [call.wibble()])
+
+
+    def test_mock_call_not_equal_leaf(self):
+        m = Mock()
+        m.foo().bar()
+        self.assertNotEqual(m.mock_calls[1], call.foo().baz())
```
Comment (mariocj89) [inline Lib/unittest/test/testmock/testmock.py]: 🤣 I've been scratching my head for 3 minutes until I noticed the `baz` rather than `bar`

File: Lib/unittest/mock.py [hunk 6]
```diff
@@ -977,46 +977,51 @@ def _mock_call(_mock_self, *args, **kwargs):
...
+            this_mock_call = _Call((mock_call_name, args, kwargs))
             _new_parent.mock_calls.append(this_mock_call)
+
+            if _new_parent._mock_new_name:
+                if is_a_call:
+                    dot = ''
+                else:
+                    dot = '.'
+                is_a_call = _new_parent._mock_new_name == '()'
+                mock_call_name = _new_parent._mock_new_name + dot + mock_call_name
```
Comment (mariocj89) [inline Lib/unittest/mock.py]: is this variable used?

File: Lib/unittest/mock.py [hunk 7]
```diff
@@ -977,46 +977,51 @@ def _mock_call(_mock_self, *args, **kwargs):
...
+            # handle mock_calls:
+            this_mock_call = _Call((mock_call_name, args, kwargs))
             _new_parent.mock_calls.append(this_mock_call)
+
+            if _new_parent._mock_new_name:
+                if is_a_call:
+                    dot = ''
+                else:
+                    dot = '.'
+                is_a_call = _new_parent._mock_new_name == '()'
```
Comment (mariocj89) [inline Lib/unittest/mock.py:1017]: Is this var used?

File: Doc/library/unittest.mock-examples.rst [hunk 8]
```diff
@@ -166,6 +166,15 @@ You use the :data:`call` object to construct lists for comparing with
     >>> mock.mock_calls == expected
     True
 
+However, parameters to calls that return mocks are not recorded, which means tracking
+nested calls where the parameters used to create ancestors are important is not possible:
```
Comment (mariocj89) [inline Doc/library/unittest.mock-examples.rst]: Not sure what this means:

> the parameters used to create ancestors are important is not possible

File: Doc/library/unittest.mock-examples.rst [hunk 9]
```diff
@@ -512,16 +521,6 @@ call::
     >>> something.backend = mock_backend
     >>> something.method()
 
-Using :attr:`~Mock.mock_calls` we can check the chained call with a single
```
Comment (mariocj89) [inline Doc/library/unittest.mock-examples.rst]: Why do you remove this? This works fine.

File: Doc/library/unittest.mock-examples.rst [hunk 10]
```diff
@@ -1144,6 +1143,15 @@ the ``mock_calls`` attribute on the manager mock:
     >>> manager.mock_calls == expected_calls
     True
 
+However, parameters to calls that return mocks are not recorded, which means tracking
+nested calls where the parameters used to create ancestors are important is not possible:
```
Comment (mariocj89) [inline Doc/library/unittest.mock-examples.rst]: Note this is possible, you just need to use `call_list()`, as it was specified in the docs.

File: Doc/library/unittest.mock.rst [hunk 11]
```diff
@@ -702,6 +702,19 @@ the *new_callable* argument to :func:`patch`.
         unpacked as tuples to get at the individual arguments. See
         :ref:`calls as tuples <calls-as-tuples>`.
 
+        .. warning::
+
+            The way :attr:`mock_calls` are recorded means that where nested
+            calls are made, the parameters of ancestor calls are not recorded
```
Comment (mariocj89) [inline Doc/library/unittest.mock.rst:708]: They are recorded, just as different calls.. :/ I think this is going to be really confusing for the reader.

**Review by mariocj89 — COMMENTED** [review]
A bunch of comments. I'd personally prefer @vstinner or @voidspace to review this PR as I don't agree with it (now, good news, I am just a *random reviewer*).

When validating calls in mocks, AFAIK people are expected to use `mock_calls` and `call_list`, that works file, see the example of the warnings you put:

```python
>>> m = Mock()
>>> m.factory(important=True).deliver()
<Mock name='mock.factory().deliver()' id='...'>
>>> m.mock_calls[-1] == call.factory(important=False).deliver()
True
```

I think this is just a wrong usage of `call`, people should just always call `call_list` if that is what they want.
This works as expected if the API is used as documetned:
```python
>>> m = Mock()
>>> m.factory(important=True).deliver()
<Mock name='mock.factory().deliver()' id='...'>
>>> m.mock_calls == call.factory(important=False).deliver().call_list()
False
```

The way `call` is done, allows you to also have the option of validating the call without caring about the parent argumetns if you wanted.

With that said about the API, I see three problems about this PR though:

1) It does not address the original concern about the usage (which I disagree with though), and with it the `call` `__eq__` "feels" different. Right now `__eq__` on a mock and a call should be done by using a method that returns you all the calls that happend (in case you want to validate all the arguments), this PR removes that option, and what is worse, people might be already using this API with that objective in their minds (which is what it seems to provde). I would **not backport this** to previous versions of Python.

2) Adds warnings about the a usage of `call` that is not the intended per the documentation.

3) Makes quite a refactor for a PR that plans to backport, I'd just "leave previous versions as they are" to minimize risks... :/ Maybe I am too paranoid.


### TL;DR;

I would just change this whole PR for a warning in the `call` class saying: 

> When used with nested `call`, make sure to use `call_list()` if you want to compare recursively with the value of the argumetns.

Real apologies if you feel I am being "hard on the PR" but I just find things are *fine*. `call(x=1).foo == call(x=2).foo` because that is how the API is. If you want to check recursively you should do: `call(x=1).foo.call_list() == call(x=2).foo.call_list()`. Anyway, as said, just a random reviewer, there are 3 core devs in this PR, one of them the original designer of this API :)

File: Lib/unittest/mock.py (hunk 4 above)
Comment (cjw296) [reply Lib/unittest/mock.py]: I don't understand what you'd like here.

File: Lib/unittest/mock.py (hunk 6 above)
Comment (cjw296) [inline Lib/unittest/mock.py:1018]: Yes.

File: Lib/unittest/mock.py (hunk 7 above)
Comment (cjw296) [reply Lib/unittest/mock.py:1017]: Yes.

File: Doc/library/unittest.mock-examples.rst (hunk 8 above)
Comment (cjw296) [reply Doc/library/unittest.mock-examples.rst]: "tracking nested calls where the parameters used to create ancestors are important is not possible"

File: Doc/library/unittest.mock-examples.rst (hunk 10 above)
Comment (cjw296) [reply Doc/library/unittest.mock-examples.rst]: Not across mocks it's not.

File: Lib/unittest/mock.py (hunk 4 above)
Comment (mariocj89) [reply Lib/unittest/mock.py]: the indentation is at the same level as the inner block. PEP8 says it should have one more indentation (4 more spaces)

File: Doc/library/unittest.mock-examples.rst (hunk 9 above)
Comment (cjw296) [reply Doc/library/unittest.mock-examples.rst]: ...because it's not safe to use: because the ancestor parameters are not compared, you can easily be fooled into thinking something is being called correctly when it is not.

File: Doc/library/unittest.mock-examples.rst (hunk 10 above)
Comment (mariocj89) [reply Doc/library/unittest.mock-examples.rst]: What do you mean with across mocks?

File: Doc/library/unittest.mock.rst (hunk 11 above)
Comment (cjw296) [reply Doc/library/unittest.mock.rst:708]: The parameters are absolutely not recorded in `mock_calls`, this is the core problem that isn't solvable.

File: Misc/NEWS.d/next/Library/2018-11-15-07-14-32.bpo-35226.wJPEEe.rst (hunk 3 above)
Comment (cjw296) [reply Misc/NEWS.d/next/Library/2018-11-15-07-14-32.bpo-35226.wJPEEe.rst]: Happy to polish this once we get some consensus on this PR.

File: Lib/unittest/mock.py (hunk 7 above)
Comment (mariocj89) [reply Lib/unittest/mock.py:1017]: indeed, I quickly overview it and it is used within the loop.

File: Doc/library/unittest.mock-examples.rst (hunk 8 above)
Comment (mariocj89) [reply Doc/library/unittest.mock-examples.rst]: OK, I see. But this can be done with `mock_calls` or `call_list` isn't it? It just cannot be all extracted from a single `call` object within `mock_calls`.

Comment (cjw296) [reply Doc/library/unittest.mock-examples.rst]: No, it cannot, if it could, I wouldn't have reported the bug, put together a PR and spent several hours trying to figure out a way to do this that was backwards compatible ;-)

File: Doc/library/unittest.mock-examples.rst (hunk 10 above)
Comment (cjw296) [reply Doc/library/unittest.mock-examples.rst]: Will cover this in a non-in-line response...

**Comment by cjw296** [comment]
@vstinner / @voidspace - sorry, I thought @mariocj89 was an owner of `unittest.mock` now, would either of you be able to look at this? I'm trying trying to get my commit bit recovered that appears to have gotten list in the move to GitHub, but regardless, I'd prefer someone else to let me know I'm not missing anything on this ;-)

My recollection of `mock_calls` being added to Mock was that it was to be able to track calls across a hierarchy of mocks. I'm fairly sure I was one of the people who requested it, although I can't find the discussion in any list archives so maybe it was in-person at a conference.

I believe the intention was always for nested calls to not compare equal if their parameters were not the same, but since this current behaviour will silently have let this incorrectly pass for many years, I'm not sure anyone would have noticed. I only spotted it after 7 years as I was looking very closely in this area, and I've been a heavy user from this feature since it arrived!

I've now put a great deal of time into thoroughly understanding the problem, and I've summarised the position above. I believe this PR represents the best that can reasonably be achieved without breaking backwards compatibility.

File: Doc/library/unittest.mock-examples.rst (hunk 9 above)
Comment (mariocj89) [reply Doc/library/unittest.mock-examples.rst]: Maybe I am misunderstanding this whole thing, but what is not safe? How can you be fooled?

just try changing foobar here to `foobaz`, the assert will fail because this is using `call_list`

File: Doc/library/unittest.mock.rst (hunk 11 above)
Comment (mariocj89) [reply Doc/library/unittest.mock.rst:708]: They are! just not in the last one (each call records only its own parameters), look:

```
>>> mock_backend.mock_calls
[call.get_endpoint('foobar'),
 call.get_endpoint().create_call('spam', 'eggs'),
 call.get_endpoint().create_call().start_call()]
```

Comment (cjw296) [reply Doc/library/unittest.mock.rst:708]: I'm not sure we're getting anywhere constructive with this, so let's wait for Michael to take a look.

File: Doc/library/unittest.mock-examples.rst (hunk 8 above)
Comment (mariocj89) [reply Doc/library/unittest.mock-examples.rst]: OK, I am clearly missing something here.

```
>>> m = Mock()
>>> m.factory(important=True).deliver()
<Mock name='mock.factory().deliver()' id='139810163105576'>
>>> m.mock_calls == call.factory(important=False).deliver().call_list()
False
```

Note the false at the end. This works AFAIK.

Setting the expectation right works:
```
>>> m.mock_calls == call.factory(important=True).deliver().call_list()
True
```

**Comment by cjw296** [comment]
Here's an example of why I think the current behaviour is dangerous, and needs fixing where possible (which turns out to be only on the `_Call` side) and warning about on the `mock_call` side:

```
>>> from unittest.mock import Mock, call
>>> from subprocess import Popen
>>> Popen = Mock(spec=Popen)
>>> p1 = Popen(['rm', '-rf', '/var'])
>>> p2 = Popen(['echo', 'hello'])
>>> p1.communicate()
<Mock name='mock().communicate()' id='4422077240'>
>>> p2.kill()
<Mock name='mock().kill()' id='4422077184'>
>>> assert Popen.mock_calls == [call(['rm', '-rf', '/var']), call(['echo', 'hello']), call(['echo', 'hello']).communicate(), call(['rm', '-rf', '/var']).kill()]
```

The intention here is to test that we're killing the `rm -rf`, hopefully before its too late, and then waiting for the `echo` to complete, but the bug that we're doing the opposite isn't picked up with this test which will pass either way. Real world issue relating to this came about from trying to fix https://github.com/Simplistix/testfixtures/issues/90.

File: Doc/library/unittest.mock.rst (hunk 11 above)
Comment (cjw296) [reply Doc/library/unittest.mock.rst:708]: ...except when `get_endpoint` is called more than once.

**Comment by vstinner** [comment]
I have been asked to review this change, but I see a lot of comments which are not marked as "resolved" and I don't see any approval, so I prefer to not review this change yet.

File: Doc/library/unittest.mock-examples.rst (hunk 8 above)
Comment (voidspace) [reply Doc/library/unittest.mock-examples.rst]: A call to any individual mock always returns the same mock (unless you're using a side_effect to return a different one). So you can't track the parent parameters in general as soon as there's more than one call. As noted the parameters to the original call will still be in the full call chain. I think the behaviour here is working as designed it just doesn't solve every use case.

File: Doc/library/unittest.mock-examples.rst (hunk 9 above)
Comment (voidspace) [reply Doc/library/unittest.mock-examples.rst]: I don't mind a note being added, although perhaps just the note without the warning. ``call_list`` is still useful so I don't want to see this removed from the docs.

File: Lib/unittest/mock.py (hunk 1 above)
Comment (voidspace) [reply Lib/unittest/mock.py:2062]: Where there's a single call there's no parent call. So having parent not set in that example looks correct.

**Review by voidspace — CHANGES_REQUESTED** [review]
The doc changes are excessively dramatic. call is still useful, but because of the standard behaviour of mocks (repeated calls will always return the same mock) parent parameters aren't tracked. However the original call is still there in the chain.

**Review by voidspace — APPROVED** [review]
LGTM
